Add initial build script for stage3_mlxupdate image

stephen-soltesz opened the pull request:
This change adds an initial stage3 image suitable for flashing a Mellanox ROM to a local NIC.

Like the other "setup" scripts, this build process can be run from a docker container.

Subsequent changes will include scripts to download and flash the ROM.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/m-lab/epoxy-images/12)
<!-- Reviewable:end -->

gfr10598 commented:
A few comments.  My understanding of this code is marginal, but AFAICT, it
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[setup_stage3_mlxupdate.sh, line 17 at r1](https://reviewable.io:443/reviews/m-lab/epoxy-images/12#-L2Qfv5o6fFuJlRmfEW9:-L2Qfv5o6fFuJlRmfEWA:br0bszd) ([raw file](https://github.com/m-lab/epoxy-images/blob/f4354ac27c1a9a0c557b662918835420f089b4a3/setup_stage3_mlxupdate.sh#L17)):*
> ```Shell
> BUILDDIR=$( realpath $BUILDDIR )
> 
> CONFIGDIR=${2:?Name of configuration}
> ```

?Directory containing configuration

---

*[setup_stage3_mlxupdate.sh, line 23 at r1](https://reviewable.io:443/reviews/m-lab/epoxy-images/12#-L2Qg8NKdA1FV--6XPZO:-L2Qg8NKdA1FV--6XPZP:bb7ygcj) ([raw file](https://github.com/m-lab/epoxy-images/blob/f4354ac27c1a9a0c557b662918835420f089b4a3/setup_stage3_mlxupdate.sh#L23)):*
> ```Shell
> BOOTSTRAP=$BUILDDIR/initramfs_${CONFIG_NAME}
> 
> function unpack () {
> ```

unpack_url

---

*[setup_stage3_mlxupdate.sh, line 36 at r1](https://reviewable.io:443/reviews/m-lab/epoxy-images/12#-L2QgS-VGvrCOBc7osd3:-L2QgS-VGvrCOBc7osd4:bza5psd) ([raw file](https://github.com/m-lab/epoxy-images/blob/f4354ac27c1a9a0c557b662918835420f089b4a3/setup_stage3_mlxupdate.sh#L36)):*
> ```Shell
> 
> 
> function enter_with_proc() {
> ```

better name for this function?  It doesn't convey to me what it does.
mount_proc_and_sys ?

---

*[setup_stage3_mlxupdate.sh, line 43 at r1](https://reviewable.io:443/reviews/m-lab/epoxy-images/12#-L2QgWTsoxCQVc9Y3phW:-L2QgWTsoxCQVc9Y3phX:b-t71nx5) ([raw file](https://github.com/m-lab/epoxy-images/blob/f4354ac27c1a9a0c557b662918835420f089b4a3/setup_stage3_mlxupdate.sh#L43)):*
> ```Shell
> 
> 
> function exit_with_proc() {
> ```

ditto
cleanup_proc_and_sys?

---

*[setup_stage3_mlxupdate.sh, line 48 at r1](https://reviewable.io:443/reviews/m-lab/epoxy-images/12#-L2Qgak1CPKJyRnublOe:-L2Qgak1CPKJyRnublOf:b-pu55nr) ([raw file](https://github.com/m-lab/epoxy-images/blob/f4354ac27c1a9a0c557b662918835420f089b4a3/setup_stage3_mlxupdate.sh#L48)):*
> ```Shell
>     umount $bootstrap/sys
> }
> ```

Maybe add a banner here to separate the mainline from the function defs?

---

*[setup_stage3_mlxupdate.sh, line 152 at r1](https://reviewable.io:443/reviews/m-lab/epoxy-images/12#-L2Qin2deQ_FxzFmK5m2:-L2Qin2deQ_FxzFmK5m3:b-bz9syu) ([raw file](https://github.com/m-lab/epoxy-images/blob/f4354ac27c1a9a0c557b662918835420f089b4a3/setup_stage3_mlxupdate.sh#L152)):*
> ```Shell
> 
> exit_with_proc $BOOTSTRAP
> ```

Should ```trap '' EXIT``` go here?

---

*[setup_stage3_mlxupdate.sh, line 159 at r1](https://reviewable.io:443/reviews/m-lab/epoxy-images/12#-L2Qj7uW7ZgFnCbSjME9:-L2Qj7uW7ZgFnCbSjMEA:b-lye2t0) ([raw file](https://github.com/m-lab/epoxy-images/blob/f4354ac27c1a9a0c557b662918835420f089b4a3/setup_stage3_mlxupdate.sh#L159)):*
> ```Shell
> # Kernel panics unless /init is defined. Use systemd for init.
> ln --force --symbolic sbin/init $BOOTSTRAP/init
> cp $CONFIGDIR/fstab       $BOOTSTRAP/etc/fstab
> ```

odd whitespace?

---

*[setup_stage3_mlxupdate.sh, line 195 at r1](https://reviewable.io:443/reviews/m-lab/epoxy-images/12#-L2QjzTrs7Yz4blBz5Mx:-L2QjzTrs7Yz4blBz5My:bbh53uk) ([raw file](https://github.com/m-lab/epoxy-images/blob/f4354ac27c1a9a0c557b662918835420f089b4a3/setup_stage3_mlxupdate.sh#L195)):*
> ```Shell
> # TODO: investigate ssh-import-id as an alternative here, or a copy from GCS.
> # echo "Adding SSH authorized keys"
> # mkdir -p $BOOTSTRAP/root/.ssh
> ```

Should we also whitelist the IP ranges allowed to ssh?

---

*[configs/stage3_mlxupdate/init, line 12 at r1](https://reviewable.io:443/reviews/m-lab/epoxy-images/12#-L2QkWhmutdEKVMvjTtH:-L2QkWhnW-qpAI4raMOH:bwr0k6o) ([raw file](https://github.com/m-lab/epoxy-images/blob/f4354ac27c1a9a0c557b662918835420f089b4a3/configs/stage3_mlxupdate/init#L12)):*
> ```
> /etc/rc.local
> 
> /sbin/dropbear
> ```

comment that this is lightweight ssh?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/epoxy-images/12#-:-L2QltAghPcRhd9Rou_u:b-k1cz0x)*
<!-- Sent from Reviewable.io -->

stephen-soltesz commented:
Review status: 6 of 7 files reviewed at latest revision, 9 unresolved discussions.

---

*[setup_stage3_mlxupdate.sh, line 17 at r1](https://reviewable.io:443/reviews/m-lab/epoxy-images/12#-L2Qfv5o6fFuJlRmfEW9:-L2QnEtEsAW4NJgiliop:b-896fix) ([raw file](https://github.com/m-lab/epoxy-images/blob/f4354ac27c1a9a0c557b662918835420f089b4a3/setup_stage3_mlxupdate.sh#L17)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

?Directory containing configuration
</blockquote></details>

Done.

---

*[setup_stage3_mlxupdate.sh, line 23 at r1](https://reviewable.io:443/reviews/m-lab/epoxy-images/12#-L2Qg8NKdA1FV--6XPZO:-L2QnIQG55DBjingILXM:b-896fix) ([raw file](https://github.com/m-lab/epoxy-images/blob/f4354ac27c1a9a0c557b662918835420f089b4a3/setup_stage3_mlxupdate.sh#L23)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

unpack_url
</blockquote></details>

Done.

---

*[setup_stage3_mlxupdate.sh, line 36 at r1](https://reviewable.io:443/reviews/m-lab/epoxy-images/12#-L2QgS-VGvrCOBc7osd3:-L2Qn_28yFCwIFHopkJH:b-896fix) ([raw file](https://github.com/m-lab/epoxy-images/blob/f4354ac27c1a9a0c557b662918835420f089b4a3/setup_stage3_mlxupdate.sh#L36)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

better name for this function?  It doesn't convey to me what it does.
mount_proc_and_sys ?

</blockquote></details>

Done.

---

*[setup_stage3_mlxupdate.sh, line 43 at r1](https://reviewable.io:443/reviews/m-lab/epoxy-images/12#-L2QgWTsoxCQVc9Y3phW:-L2Qnc7Xa0VNI73-OhLY:bdm87y6) ([raw file](https://github.com/m-lab/epoxy-images/blob/f4354ac27c1a9a0c557b662918835420f089b4a3/setup_stage3_mlxupdate.sh#L43)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

ditto
cleanup_proc_and_sys?

</blockquote></details>

I was trying to mirror the behavior of `pushd` and `popd`. But, they don't quite work that way so it's a poor metaphor. I've updated the function names to be more representative of what they're doing. 

`umount_proc_and_sys` here.

---

*[setup_stage3_mlxupdate.sh, line 48 at r1](https://reviewable.io:443/reviews/m-lab/epoxy-images/12#-L2Qgak1CPKJyRnublOe:-L2Qo08joITEIMAPJb1s:b-896fix) ([raw file](https://github.com/m-lab/epoxy-images/blob/f4354ac27c1a9a0c557b662918835420f089b4a3/setup_stage3_mlxupdate.sh#L48)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Maybe add a banner here to separate the mainline from the function defs?
</blockquote></details>

Done.

---

*[setup_stage3_mlxupdate.sh, line 152 at r1](https://reviewable.io:443/reviews/m-lab/epoxy-images/12#-L2Qin2deQ_FxzFmK5m2:-L2Qo8Bg2h-ZLeOdY0QI:b-duzahh) ([raw file](https://github.com/m-lab/epoxy-images/blob/f4354ac27c1a9a0c557b662918835420f089b4a3/setup_stage3_mlxupdate.sh#L152)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Should ```trap '' EXIT``` go here?
</blockquote></details>

Yes! Done.

---

*[setup_stage3_mlxupdate.sh, line 159 at r1](https://reviewable.io:443/reviews/m-lab/epoxy-images/12#-L2Qj7uW7ZgFnCbSjME9:-L2QoJ2_YeKd3z0wS5EA:b-896fix) ([raw file](https://github.com/m-lab/epoxy-images/blob/f4354ac27c1a9a0c557b662918835420f089b4a3/setup_stage3_mlxupdate.sh#L159)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

odd whitespace?
</blockquote></details>

Done.

---

*[setup_stage3_mlxupdate.sh, line 195 at r1](https://reviewable.io:443/reviews/m-lab/epoxy-images/12#-L2QjzTrs7Yz4blBz5Mx:-L2QoN1PzBmAzuaIbpIl:bthwze5) ([raw file](https://github.com/m-lab/epoxy-images/blob/f4354ac27c1a9a0c557b662918835420f089b4a3/setup_stage3_mlxupdate.sh#L195)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Should we also whitelist the IP ranges allowed to ssh?
</blockquote></details>

I don't think so. But, happy to talk more in person.

---

*[configs/stage3_mlxupdate/init, line 12 at r1](https://reviewable.io:443/reviews/m-lab/epoxy-images/12#-L2QkWhmutdEKVMvjTtH:-L2QqblrmE-ZRUskdyfw:b-flhjzw) ([raw file](https://github.com/m-lab/epoxy-images/blob/f4354ac27c1a9a0c557b662918835420f089b4a3/configs/stage3_mlxupdate/init#L12)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

comment that this is lightweight ssh?
</blockquote></details>

Ah, this entire file is unnecessary! Removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/epoxy-images/12)*
<!-- Sent from Reviewable.io -->
